packages/nimble-components/src/table-column/number-text/models/unit-scale-formatter.ts

```diff
@@ -0,0 +1,91 @@
+import type { ScaledUnit } from './scaled-unit';
+import { NumberFormatter } from './number-formatter';
+
+export type UnitScaleFormatterConstructor = new (
+    lang: string,
+    formatterOptions: Intl.NumberFormatOptions
+) => UnitScaleFormatter;
+
+/**
+ * A formatter for a number-text column that supports unit labels.
+ */
+export abstract class UnitScaleFormatter extends NumberFormatter {
+    public alwaysUseBaseUnit = false;
+    private _supportedUnits?: ScaledUnit[];
+    private _baseUnit?: ScaledUnit;
+
+    private get supportedUnits(): ScaledUnit[] {
+        if (this._supportedUnits === undefined) {
+            this.setSupportedUnitsAndBaseUnit();
+        }
+        return this._supportedUnits!;
+    }
+
+    private get baseUnit(): ScaledUnit {
+        if (this._baseUnit === undefined) {
+            this.setSupportedUnitsAndBaseUnit();
+        }
+        return this._baseUnit!;
+    }
+
+    public constructor(
+        private readonly lang: string,
+        private readonly formatterOptions: Intl.NumberFormatOptions
+    ) {
+        super();
+    }
+
+    public getValueForBestUnit(number: number): number {
+        const unit = this.pickBestUnit(number);
+        return number / unit.conversionFactor;
+    }
+
+    protected abstract getSupportedUnits(
+        lang: string,
+        formatterOptions: Intl.NumberFormatOptions
+    ): ScaledUnit[];
+
+    protected override format(number: number): string {
+        const unit = this.pickBestUnit(number);
+        return unit.format(number / unit.conversionFactor);
+    }
+
+    private pickBestUnit(number: number): ScaledUnit {
+        const magnitude = Math.abs(number);
+        if (
+            magnitude === 0
+            || magnitude === Infinity
+            || Number.isNaN(magnitude)
+            || this.alwaysUseBaseUnit
+        ) {
+            return this.baseUnit;
+        }
+        for (const unit of this.supportedUnits) {
+            if (magnitude >= unit.conversionFactor) {
+                return unit;
+            }
+        }
+        return this.supportedUnits[this.supportedUnits.length - 1]!;
```

Comment:
I'm not sure it's worth it when we'll be dealing with arrays on the order of a dozen elements (especially since there's apparently no built-in JS implementation), but I guess we can. We do run `pickBestUnit` twice for each cell, so maybe that's enough reason.

Reply:
yea... you're right, with small sets the optimizations may not always pan out. May not be worth it unless we measure.

We should probably have a test with a large grid of numbers like at least 10x10 visible and rendering (if not much more like excel spreadsheet level) and make sure there are no regressions and that there aren't any specific bottlenecks happening.

Reply:
Do you mean an automated performance test? That would be our first one, and I'm not sure how we would capture the rendering time. Or were you suggesting a manual test via a Storybook story or addition to the example app?

More generally, we probably need to have some performance-based acceptance testing of the table and all its various features at some point. Not sure if that would be automated, manual, or a mix.

Reply:
Example app is not a test bed so that's not the right spot.
A story could be possible.

My specific concern with this PR is avoiding regressions to numeric performance as is. Want to make sure if you are not using units that we aren't paying extra overhead. Think the general concern is separate and true.

Reply:
I added a new story that is a 30x20 table of numbers. We can check the responsiveness of this table for obvious regressions when we make significant changes to the table or number column. To compare against pre-unit performance, I added the same story in my local `main` branch. The good news is there's no percievable difference. The bad news is that even before my change, upon initial load, the table and the storybook controls (i.e. unit selection radio buttons) are unresponsive for about five seconds. When I pause execution during the unresponsive period, the call stack is in `axe.js`, so my guess is that Storybook's accessibility plug-in is to blame.

Reply:
> When I pause execution during the unresponsive period, the call stack is in axe.js, so my guess is that Storybook's accessibility plug-in is to blame.

The axe plugin can be disabled for stories that create a lot of DOM, see matrix test configs: https://github.com/ni/nimble/blob/a844a186ce955d3f25ea1b58f469f70ff378916a/packages/nimble-components/src/utilities/tests/matrix.ts#L7

> The good news is there's no perceivable difference.

I'd like a stronger indication that there is no change in performance with and without the change. The described test doesn't give much confidence for that.

One idea is to add a mode to the story that updates the data set on a short interval. That way when running a performance profile the initial load is not captured and multiple executions can be sampled. I'd do something that pregenerate a couple data sets that can be cycled between.

Also open to other ideas to give a more confident measurement of that rendering perf.

Reply:
I added a play function to the story that swaps data every second. Unfortunately, I couldn't use a story argument to turn this on/off, because I couldn't get the current argument values in the play function. Anyway, I copied the same changes to main for comparision. It looks like both with and without my changes, it takes about 25 ms to render. (The whole 30x20 table was visible.)
![image](https://github.com/ni/nimble/assets/7282195/05b4a853-4a34-45cb-aea7-dcd111345cf5)

Reply:
> The axe plugin can be disabled for stories that create a lot of DOM, see matrix test configs:

It turned out that we weren't actually disabling the a11y plugin anywhere, because we were setting `disabled` rather than `disable`. I fixed those.

Reply:
Thanks for putting that together! That layout pass is pretty darn long, almost a full frame of time to calculate layout is surprising but not introduced by this change so looks good 👍

Reply:
In order to reliably handle the rounding effects of all possible `Intl.NumberFormat` options (as opposed to just `maximumFractionDigits`), I had to make significant changes involving additional calls to `Intl.NumberFormat.format()`. My initial, straightforward approach increased the render time from ~25ms to ~35ms. I felt that was too much, so I optimized the algorithm and got render times back down to ~27ms:
![image](https://github.com/ni/nimble/assets/7282195/609c36c3-24e7-421a-b7b1-3ec2bf9617a6)